File: sdks/java/src/main/java/org/byteveda/taskito/proxies/ProxySession.java
package org.byteveda.taskito.proxies;

import java.lang.System.Logger;
import java.lang.System.Logger.Level;
import java.time.Duration;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.IdentityHashMap;
import java.util.Map;
import org.byteveda.taskito.errors.ProxyException;

/**
 * A unit-of-work wrapper over {@link Proxies} adding identity dedup and a
 * cleanup lifecycle. Within one session:
 *
 * <ul>
 *   <li>{@link #deconstruct(Object)} memoizes by (instance identity, purpose) —
 *       deconstructing the same object again returns the same {@link ProxyRef}
 *       without calling the handler twice. The first call's TTL wins per
 *       (instance, purpose); use a new session to refresh expiry.
 *   <li>{@link #reconstruct(ProxyRef)} memoizes by the ref's signature — every
 *       ref to the same underlying resource resolves to the same instance, and
 *       the handler reconstructs it once. Signature, expiry, and purpose are
 *       re-verified on every call, memo hit or not.
 *   <li>{@link #close()} runs {@link ProxyHandler#cleanup} once per unique
 *       reconstructed instance, in reverse reconstruction order (LIFO).
 * </ul>
 *
 * <p>Not thread-safe — confine a session to the thread that created it. A
 * session models one producer batch or one task invocation.
 */
public final class ProxySession implements AutoCloseable {
    private static final Logger LOG = System.getLogger(ProxySession.class.getName());

    private final Proxies proxies;
    private final IdentityHashMap<Object, Map<String, ProxyRef>> deconstructed = new IdentityHashMap<>();
    private final Map<String, Object> reconstructed = new HashMap<>();
    private final Deque<Runnable> cleanups = new ArrayDeque<>();
    private boolean closed;

    ProxySession(Proxies proxies) {
        this.proxies = proxies;
    }

    /** Deconstruct {@code value} (no expiry or purpose), deduped by instance identity. */
    public ProxyRef deconstruct(Object value) {
        return deconstruct(value, null, null);
    }

    /** Deconstruct {@code value} with a TTL, deduped by instance identity. */
    public ProxyRef deconstruct(Object value, Duration ttl) {
        return deconstruct(value, ttl, null);
    }

    /**
     * Deconstruct {@code value} bound to {@code ttl}/{@code purpose} (both
     * nullable), deduped by (instance identity, purpose).
     */
    public ProxyRef deconstruct(Object value, Duration ttl, String purpose) {
        ensureOpen();
        if (value == null) {
            throw new ProxyException("cannot deconstruct null");
        }
        Map<String, ProxyRef> byPurpose = deconstructed.computeIfAbsent(value, key -> new HashMap<>());
        ProxyRef cached = byPurpose.get(purpose);
        if (cached != null) {
            return cached;
        }
        ProxyRef ref = proxies.deconstruct(value, ttl, purpose);
        byPurpose.put(purpose, ref);
        return ref;
    }

    /** Verify and reconstruct {@code ref}, deduped by its signature. */
    public Object reconstruct(ProxyRef ref) {
        return reconstruct(ref, null);
    }

    /**
     * Verify (always — including memo hits, so a ref that expires mid-session
     * stops resolving) and reconstruct {@code ref}, deduped by its signature.
     */
    public Object reconstruct(ProxyRef ref, String expectedPurpose) {
        ensureOpen();
        ProxyHandler<Object> handler = proxies.handlerFor(ref.handler());
        proxies.verify(ref, expectedPurpose);
        String signature = ref.signature();
        if (reconstructed.containsKey(signature)) {
            return reconstructed.get(signature);
        }
        Object value = handler.reconstruct(ref.reference());
        reconstructed.put(signature, value);
        cleanups.push(() -> handler.cleanup(value));
        return value;
    }

    /** {@link #reconstruct(ProxyRef)} cast to the caller's type. */
    @SuppressWarnings("unchecked")
    public <T> T resolve(ProxyRef ref) {
        return (T) reconstruct(ref);
    }

    /** {@link #reconstruct(ProxyRef, String)} cast to the caller's type. */
    @SuppressWarnings("unchecked")
    public <T> T resolve(ProxyRef ref, String expectedPurpose) {
        return (T) reconstruct(ref, expectedPurpose);
    }

    /**
     * Run {@link ProxyHandler#cleanup} for every reconstructed instance in
     * reverse order (LIFO), once each. A cleanup failure is logged and never
     * skips the rest. Idempotent.
     */
    @Override
    public void close() {
        if (closed) {
            return;
        }
        closed = true;
        while (!cleanups.isEmpty()) {
            Runnable cleanup = cleanups.pop();
            try {
                cleanup.run();
            } catch (RuntimeException e) {
                // Cleanup must never fail the rest of the teardown — record and continue.
                LOG.log(Level.WARNING, "proxy cleanup failed", e);
            }
        }
        deconstructed.clear();
        reconstructed.clear();
    }

    private void ensureOpen() {
        if (closed) {
            throw new ProxyException("proxy session is closed");
        }
    }
}

File: sdks/node/src/proxies/proxy-session.ts
import { ProxyError } from "../errors";
import { createLogger } from "../utils/logger";
import type { Proxies } from "./proxies";
import type { ProxyRef } from "./types";

const log = createLogger("proxies");

/**
 * A unit-of-work wrapper over {@link Proxies} adding identity dedup and a
 * cleanup lifecycle. Within one session:
 *
 * - {@link ProxySession.deconstruct} memoizes by (instance identity, purpose)
 *   — deconstructing the same object again returns the same {@link ProxyRef}
 *   without calling the handler twice. The first call's TTL wins per
 *   (instance, purpose); use a new session to refresh expiry.
 * - {@link ProxySession.reconstruct} memoizes by the ref's signature — every
 *   ref to the same underlying resource resolves to the same instance, and
 *   the handler reconstructs it once. Signature, expiry, and purpose are
 *   re-verified on every call, memo hit or not, so a ref that expires
 *   mid-session stops resolving.
 * - {@link ProxySession.close} runs the handler's `cleanup` once per unique
 *   reconstructed instance, in reverse reconstruction order (LIFO).
 *
 * Not thread-safe conceptually — a session models one producer batch or one
 * task invocation; don't share it across concurrent work.
 */
export class ProxySession {
  private readonly proxies: Proxies;
  private readonly deconstructed = new Map<unknown, Map<string | undefined, ProxyRef>>();
  private readonly reconstructed = new Map<string, unknown>();
  private readonly cleanups: Array<() => void> = [];
  private closed = false;

  /** Construct via {@link Proxies.session}, not directly. */
  constructor(proxies: Proxies) {
    this.proxies = proxies;
  }

  /**
   * Deconstruct `value` (optionally bound to a TTL and a purpose), deduped by
   * (instance identity, purpose). The underlying registry deconstructs each
   * (instance, purpose) pair once; a later call with a different `ttlMs` is
   * silently ignored — open a new session to refresh expiry.
   */
  deconstruct(value: unknown, opts?: { ttlMs?: number; purpose?: string }): ProxyRef {
    this.ensureOpen();
    const purpose = opts?.purpose;
    const cached = this.deconstructed.get(value)?.get(purpose);
    if (cached) {
      return cached;
    }
    // Delegate before memoizing so an invalid value (e.g. null) throws the
    // registry's error without leaving a memo entry behind.
    const ref = this.proxies.deconstruct(value, opts);
    let byPurpose = this.deconstructed.get(value);
    if (!byPurpose) {
      byPurpose = new Map();
      this.deconstructed.set(value, byPurpose);
    }
    byPurpose.set(purpose, ref);
    return ref;
  }

  /**
   * Verify (always — including memo hits, so a ref that expires mid-session
   * stops resolving) and reconstruct `ref`, deduped by its signature.
   */
  reconstruct(ref: ProxyRef, expectedPurpose?: string): unknown {
    this.ensureOpen();
    const handler = this.proxies.handlerFor(ref.handler);
    this.proxies.verifyRef(ref, expectedPurpose);
    if (this.reconstructed.has(ref.signature)) {
      return this.reconstructed.get(ref.signature);
    }
    const value = handler.reconstruct(ref.reference);
    this.reconstructed.set(ref.signature, value);
    this.cleanups.push(() => handler.cleanup?.(value));
    return value;
  }

  /** {@link ProxySession.reconstruct} cast to the caller's type. */
  resolve<T>(ref: ProxyRef, expectedPurpose?: string): T {
    return this.reconstruct(ref, expectedPurpose) as T;
  }

  /**
   * Run the handler's `cleanup` for every reconstructed instance in reverse
   * order (LIFO), once each. Cleanup failures are logged and never skip the
   * rest — close always completes and never throws (teardown convention).
   * Idempotent.
   */
  close(): void {
    if (this.closed) {
      return;
    }
    this.closed = true;
    for (let cleanup = this.cleanups.pop(); cleanup !== undefined; cleanup = this.cleanups.pop()) {
      try {
        cleanup();
      } catch (error) {
        // Cleanup must never fail the rest of the teardown — log and continue.
        log.warn("proxy cleanup failed", error);
      }
    }
    this.deconstructed.clear();
    this.reconstructed.clear();
  }

  [Symbol.dispose](): void {
    this.close();
  }

  private ensureOpen(): void {
    if (this.closed) {
      throw new ProxyError("proxy session is closed");
    }
  }
}
